contracts/utils/EnumerableSet.sol

```diff
@@ -0,0 +1,84 @@
+pragma solidity ^0.5.10;
+
+
+/**
+ * @title EnumerableSet
+ * @dev Data structure - https://en.wikipedia.org/wiki/Set_(abstract_data_type)
+ *
+ * An EnumerableSet.AddressSet is a data structure containing a number of unique addresses.
+ *
+ *  - It is possible to add and remove addresses in O(1).
+ *  - It is also possible to query if the AddressSet contains an address in O(1).
+ *  - It is possible to retrieve an array with all the addresses in the AddressSet using enumerate.
+ *    This operation is O(N) where N is the number of addresses in the AddressSet. The order in
+ *    which the addresses are retrieved is not guaranteed.
+ *
+ * Initialization of a set must include an empty array:
+ * `EnumerableSet.AddressSet set = EnumerableSet.AddressSet({values: new address[](0)});`
+ *
+ * @author Alberto Cuesta Cañada
+ */
+library EnumerableSet {
+
+    event ValueAdded(address value);
+    event ValueRemoved(address value);
+
+    struct AddressSet {
+        // Position of the value in the `values` array, plus 1 because index 0
+        // means a value is not in the set.
+        mapping (address => uint256) index;
+        address[] values;
+    }
+
+    /**
+     * @dev Add a value. O(1).
+     */
+    function add(AddressSet storage set, address value)
+        internal
+    {
+        if (!contains(set, value)){
+            set.index[value] = set.values.push(value);
+            emit ValueAdded(value);
+        }
+    }
+
+    /**
+     * @dev Remove a value. O(1).
+     */
+    function remove(AddressSet storage set, address value)
+        internal
+    {
+        if (contains(set, value)) {
+            set.values[set.index[value] - 1] = set.values[set.values.length - 1];
```

Comment:
Could you add a short comment here explaining the swap-and-pop operation? The code by itself is a bit cryptic, I think that would greatly improve readability.

Reply:
Yes, the multiple `set.` and having to remove one from the index don't help.

I added the comment.

I can also use temporary variables to make it clearer, or create an `index(set, value)` function that returns `set.index[value] - 1`, which would make clear that the indexes are offset by one.